clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go

```diff
@@ -0,0 +1,284 @@
+	}
+}
+
+type LoadBalancerNodeSyncMeasurement struct {
+	client clientset.Interface
+	// selector used to select relevant load balancer type service used for measurement
+	selector *measurementutil.ObjectSelector
+	// waitTimeout specify for the timeout for node sync on all LBs to complete
+	waitTimeout time.Duration
+	// svcNodeSyncLatencyTracker tracks the nodesync latency
+	svcNodeSyncLatencyTracker *measurementutil.ObjectTransitionTimes
+	// excludedNodeName is the node name used to trigger LB nodesync
+	excludedNodeName string
+}
+
+// LoadBalancerNodeSyncMeasurement takes measurement of node sync latency for selected lb type services.
+// This measurement only works for K8s 1.19 as it depends on the ExcludeNodeForLoadbalancer label.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all LoadBalancer type service with all-namespace scope is assumed.
+// "measure" action triggers nodesync and observation of nodesync completion for selected LB services.
+// "gather" returns node sync latency summary.
+func (s *LoadBalancerNodeSyncMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "measure":
+		if err := s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultNodeSyncLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.measureNodeSyncLatency()
+	case "gather":
+		if err := s.labelNodeForLBs(false); err != nil {
+			return nil, err
+		}
+		return s.gather(config.Identifier)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) Dispose() {}
+
+func (s *LoadBalancerNodeSyncMeasurement) String() string {
+	return loadBalancerNodeSyncLatencyName + ": " + s.selector.String()
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) measureNodeSyncLatency() error {
+	ctx := context.Background()
+	options := metav1.ListOptions{}
+	s.selector.ApplySelectors(&options)
+	svcList, err := s.client.CoreV1().Services(s.selector.Namespace).List(ctx, options)
+	if err != nil {
+		return err
+	}
+
+	// lbSvcMap contains key->service mapping for load balancer type services
+	lbSvcMap := map[string]v1.Service{}
+	for _, svc := range svcList.Items {
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+			lbSvcMap[keyFunc(svc.Namespace, svc.Name)] = svc
+		}
+	}
+	totalLbSvc := len(lbSvcMap)
+
+	// Use event informer to keep track of nodeSync events.
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	eventInformer := informerv1.NewEventInformer(s.client, s.selector.Namespace, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
```

Comment:
Periodic listing of all events could be problematic in the biggest clusters. Could you use approach from [ooms_tracker.go#L125](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/ooms_tracker.go#L125) to ignore relists?
It basically ignores initial listing (and listing on e.g. broken watch) and uses only watches with [field selector](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/ooms_tracker.go#L203)

Reply:
Fixed by using this pattern.